KS-4-5 задача 5

VVShchegolev opened the pull request:
Что сделано:
Были созданы константы, чтобы избежать магических чисел:
Также перед строками ввода были добавлены println(), в которых содержится информация что нужно сделать;
Также создал три переменных с условиями для удобства чтения, чтобы не растягивать код;
На счет readLine(), их можно было заменить на readln(), честно говоря не понял в чём разница между ними.

VVShchegolev committed:
KS-4-5 задача 5

ievetrovdev commented on utils/src/main/kotlin/Lesson_4_Ex_5/Lesson_4_Ex_5.kt:
```diff
@@ -0,0 +1,25 @@
+fun main() {
+
+    println("Введите наличие повреждений корабля (true/false):")
+    val isShipDamage = readLine()!!.toBoolean()
```
Использование оператора `!!` подразумевает, что ты уверен в том, что значение не может быть `null`. Однако, если это не так, то приложение "упадет" с ошибкой. Поэтому этот оператор считается опасным и не рекомендуется использовать. Только в крайних случаях. 

Это противоречит основной идее безопасности типов в Kotlin, поэтому лучше использовать альтернативные способы обработки нулабельных значений. Такие как оператор безопасного вызова `?.`, элвис-оператор `?:` или оператор `let`.

Но пока я рекомендую как раз использовать более свежую функцию `readln()`. Метод под капотом обрабатывает нулябельность и все кастует в строку (в том числе `null` становится строкой `“null”`.

ievetrovdev commented on utils/src/main/kotlin/Lesson_4_Ex_5/Lesson_4_Ex_5.kt:
```diff
@@ -0,0 +1,25 @@
+    val isWeatherFavorable = readLine()!!.toBoolean()
+
+    val isMainScenarioMet = !isShipDamage && crewSize >= MIN_CREW_SIZE && crewSize <= MAX_CREW_SIZE && provisionBoxes > MIN_PROVISION_BOXES
+    val isAlternativeScenarioMet = isShipDamage && crewSize == REQUIRED_CREW_SIZE_FOR_DAMAGE && isWeatherFavorable && provisionBoxes >= MIN_PROVISION_BOXES
```
Длинные строки получились, для читаемости лучше разбивать на несколько строк после оператора && или ||.

VVShchegolev committed:
Update KS-4-5 Задание 5

Что исправлено: 
поменял все readLine на readln;
также сделал условия в строчку для удобства чтения

ievetrovdev commented on utils/src/main/kotlin/Lesson_4_Ex_5/Lesson_4_Ex_5.kt:
```diff
@@ -0,0 +1,32 @@
+                                  crewSize == REQUIRED_CREW_SIZE_FOR_DAMAGE &&
+                                  isWeatherFavorable &&
+                                  provisionBoxes >= MIN_PROVISION_BOXES
+    val isReadyToSail = isMainScenarioMet ||
```
Эта строка не такая уж длинная, чтобы делать перенос)

ievetrovdev reviewed (approved):
Отлично!
